File: 프로그래머스/1/340213. ［PCCP 기출문제］ 1번 ／ 동영상 재생기/［PCCP 기출문제］ 1번 ／ 동영상 재생기.js
function solution(video_len, pos, op_start, op_end, commands) {
    var answer = '';
    
    var video_len_m = parseInt(video_len.substring(0, 2));
    var video_len_s = parseInt(video_len.substring(3, 5));
    var pos_m = parseInt(pos.substring(0, 2));
    var pos_s = parseInt(pos.substring(3, 5));
    var op_start_m = parseInt(op_start.substring(0, 2));
    var op_start_s = parseInt(op_start.substring(3, 5));
    var op_end_m = parseInt(op_end.substring(0, 2));
    var op_end_s = parseInt(op_end.substring(3, 5));
    
    // 시작 위치가 오프닝 사이에 존재하는 경우
    if (is_op(op_start_m, op_start_s, op_end_m, op_end_s, pos_m, pos_s)){
        pos_m = op_end_m;
        pos_s = op_end_s;
    }
    
    for (let i = 0; i < commands.length; i++){
        if (commands[i] === "next"){
            pos_s += 10;
            // 60초 이상일 경우
            if (pos_s >= 60){
                pos_m++;
                pos_s -= 60;
            }
            // 영상의 끝을 넘은 경우
            if (pos_m > video_len_m || (pos_m == video_len_m && pos_s > video_len_s)){
                pos_m = video_len_m;
                pos_s = video_len_s;
            }
            // 오프닝 사이에 존재하는 경우
            if (is_op(op_start_m, op_start_s, op_end_m, op_end_s, pos_m, pos_s)){
                pos_m = op_end_m;
                pos_s = op_end_s;
            }
        }
        else{
            pos_s -= 10;
            // 초가 음수인 경우
            if (pos_s < 0){
                pos_m--;
                pos_s += 60;
            }
            // 분이 음수인 경우
            if (pos_m < 0){
                pos_m = 0;
                pos_s = 0;
            }
            // 오프닝 사이에 존재하는 경우
            if (is_op(op_start_m, op_start_s, op_end_m, op_end_s, pos_m, pos_s)){
                pos_m = op_end_m;
                pos_s = op_end_s;
            }
        }
    }
    
    str_pos_m = String(pos_m);
    str_pos_s = String(pos_s);
    if (str_pos_m.length === 1){
        str_pos_m = "0"+str_pos_m;
    }
    if (str_pos_s.length === 1){
        str_pos_s = "0"+str_pos_s;
    }
    
    answer = str_pos_m + ":" + str_pos_s;
    
    return answer;
}

function is_op(op_start_m, op_start_s, op_end_m, op_end_s, pos_m, pos_s){
    if ((op_start_m < pos_m || (op_start_m == pos_m && op_start_s <= pos_s)) &&
       (op_end_m > pos_m || (op_end_m == pos_m && op_end_s >= pos_s))){
        return true;
    }
    return false;
}

File: 프로그래머스/1/340213. ［PCCP 기출문제］ 1번 ／ 동영상 재생기/［PCCP 기출문제］ 1번 ／ 동영상 재생기.cpp
#include <string>
#include <vector>

using namespace std;

struct Time{
    int minute;
    int second;
    Time(string pos){
        minute = stoi(pos.substr(0, 2));
        second = stoi(pos.substr(3, 2));
    }
};

string timeToStr(Time t){
    string sm = to_string(t.minute);
    if (sm.length() == 1){
        sm = "0"+sm;
    }
    string ss = to_string(t.second);
    if (ss.length() == 1){
        ss = "0"+ss;
    }
    
    string rtn = sm + ":" + ss;
    
    return rtn;
}

bool isOp(string op_start, string op_end, string pos){
    Time opStartT = Time(op_start);
    Time opEndT = Time(op_end);
    Time posT = Time(pos);
    
    int opStartS = 60 * opStartT.minute + opStartT.second;
    int opEndS = 60 * opEndT.minute + opEndT.second;
    int posS = 60 * posT.minute + posT.second;
    
    if (opStartS <= posS && posS <= opEndS){ 
        return true;
    }
    return false;
}

string cmdNext(string pos, string video_len, string op_start, string op_end){
    Time posT = Time(pos);
    posT.second += 10;
    
    while (posT.second >= 60){
        posT.second -= 60;
        posT.minute++;
    }
    if (isOp(op_start, op_end, timeToStr(posT))){
        return op_end;
    }
    Time endT = Time(video_len);
    int endS = endT.minute * 60 + endT.second;
    int posS = posT.minute * 60 + posT.second;
    
    if (posS >= endS){
        return video_len;
    }
    
    return timeToStr(posT);
}

string cmdPrev(string pos, string video_len, string op_start, string op_end){
    Time posT = Time(pos);
    posT.second -= 10;
    
    while (posT.second < 0){
        posT.second += 60;
        posT.minute--;
    }
    if (posT.minute < 0){
        posT = Time("00:00");
    }
    if (isOp(op_start, op_end, timeToStr(posT))){
        return op_end;
    }
    
    return timeToStr(posT);
}

string solution(string video_len, string pos, string op_start, string op_end, vector<string> commands) {
    string answer = "";
    
    if (isOp(op_start, op_end, pos)){
        pos = op_end;
    }
    
    for (string cmd: commands){
        if (cmd == "next"){
            pos = cmdNext(pos, video_len, op_start, op_end);
        }
        else{
            pos = cmdPrev(pos, video_len, op_start, op_end);
        }
    }
    
    answer = pos;
    return answer;
}